M2/Macaulay2/packages/KGAlgebras.m2

```diff
@@ -3,132 +3,115 @@ newPackage(
+    n := #matrixElts-1; -- maximum index, since we are starting at 0
+    groupIndices := {0..n}; -- indices
+    g := symbol g;
+    groupElts := g_0 .. g_n; -- group elements
+    FA := freeAlgebra(K,groupElts);
+    use FA; -- make FA our working ring
+    multTableIdealGens := new MutableList;
+    for i in 0 .. n do(
+        for j in 0 .. n do( 
+            productInv := position(matrixElts,M -> M==matrixElts#i * matrixElts#j);
+            H := (FA_i)*(FA_j)-(FA_productInv);
+            multTableIdealGens##multTableIdealGens = H;
+        );
+    );
+    print multTableIdealGens;
+    FA/ideal(toList multTableIdealGens) -- output
 );
 random(FreeAlgebraQuotient) := opt -> (S) -> (       
-	randelt := sum(apply(toList(0 .. #(gens S)-1), i -> random(coefficientRing S) * S_i))
+    randelt := sum(apply(toList(0 .. #(gens S)-1), i -> random(coefficientRing S) * S_i))
 );
---True implies irreducible, false implies reducible
+--True implies irreducible. Otherwise returns proper K[G]-submodule. 
 meataxe = method()
 meataxe(FiniteGroupAction, FreeAlgebraQuotient,Ring) := (G,S,K) -> (
-	local p;
-	local L;
-	local M;
-	local MP;
-	local y;
-	local m;
-	local q;
-	local Mp;
-	local M1;
-	local MPP;
-	local H;
-	local N;
-	local n;
-	local M2;
-	local NP;
-	local yp;	
-	while true do(
-		--Do the meataxe algorithm 
-		p = random S; --generate random element of KG algebra
-		L = toList coefficients p;  
-		M = (group G)#0-(group G)#0; -- zero matrix which has the same dimensions as an element of G
-		m = 0;
-		for x in (flatten entries L#0) do(
-			n = index(x);
-			M = M + (group G)#n * sub((flatten entries L#1)#m,K);
-			m= m+1	
-		); --This for loop generates a matrix representing p using the matrix elements for G
-		--Factors of the characteristic polynomial of M
-		q=apply(factor characteristicPolynomial M, y-> value y); --Get an array containing the factors of the characteristic polynomial of the matrix M
-		q=sort toList q; --Sorts q by degree
-		if (degree q#0)#0 == 0 then(
-			q=drop(q,1)--Drops constant factor if possible
-		);
-		
-		Mp = (group G)#0-(group G)#0; --zero matrix same dimensions as elements of G
-		
-		for P in q do( 
-			n = (degree P)#0; 
-			MP= Mp = (group G)#0-(group G)#0; 
-			
-			--Evaluates P on the matrix M
-			for m in 0 .. #apply(apply((entries (coefficients P)#0)#0,degree),x-> x#0)-1 do(
-				Mp=M^((apply(apply((entries (coefficients P)#0)#0,degree),x-> x#0))#m);
-				Mp=Mp*(apply((entries (coefficients P)#1),x-> x#0))#m;
-				MP = MP+Mp
-			);
-			
+    while true do(
+        --Do the meataxe algorithm 
+        p := random S; --generate random element of KG algebra
+        L := toList coefficients p;  
+        M := (group G)#0-(group G)#0; -- zero matrix which has the same dimensions as an element of G
+        m := 0;
+        for x in (flatten entries L#0) do(
+            n := index(x);
+            M = M + (group G)#n * sub((flatten entries L#1)#m,K);
+            m = m+1    
```

Comment:
This is totally optional, but if you're interested, you can use [augmented assignment](https://macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_augmented_spassignment.html) here:

```m2
M += (group G)#n * ...
m += 1
```

It also works for other binary operators like `*`, so line 74 could become:

```m2
Mp *= (apply(...
```